[TLLM-6777][feature] Support SWA KV cache reuse OOW block detach

eopXD opened the pull request:
<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description
<!--
Please explain the issue and the solution in short.
-->

This MR is a continuation of #6768. In the previous merge request, OOW (out-of-window) blocks are only detached when reuse is not enabled, that is, the block movement behavior is identical between SWA and full attention when reuse is enabled.

This merge request attempts to enable OOW block detach when reuse is enabled. The required changes are:
- Let KV cache manager keep track of which block is used by which sequence
- Remove restriction for the eviction policy to be able to release a non-leaf block

Along with the development, bugs inside `freeChildren` and offload mechanism under `getFreeBlock` is resolved because they will affect the functionality this merge request is trying to achieve.

When a block goes OOW, it is released from the sequence, it will be available to be reclaimed and the block is held by the eviction policy for another sequence to acquire upon calling. On the other hand, we want to potentially store the sequence for reuse. To safely achieve this, the record of block ownership is done under `WindowBlockManager::getFreeBlock`. If the block acquired was originally owned by another sequence that is live inside the manager, then we invalidate the sequence for store for reuse.

At the end of a sequence (when `removeSequence` is called toward it), the KV cache manager will check if the sequence has all blocks not reclaimed by another sequence. If so, then the sequence is safe to be stored for reuse and store for reuse action will be performed.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

Regarding unit test, `KVCacheManagerVariableWindowAttentionWithReuseTest`, `KVCacheManagerSWAInvalidateReuseTest`, and `KVCacheManagerMaxAttentionWindowWithReuseTest` are added to verify SWA reuse and block invalidating mechanism.

Regarding end-to-end, the existing test for vswa setting for gemma3 for the mmlu and gsm8k benchmark will cover the code paths added in this merge request.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

eopXD committed:
[kv cache manager] Implement the detach mechanism for SWA reuse

This MR is a continuation of #6768. In the previous merge request, OOW
(out-of-window) blocks are only detached when reuse is not enabled.
This MR enables KV cache manager to detach the block when reuse is
enabled.

When reuse is enabled, OOW block gets detached while the manager bookkeep
whether the block is overwritten by another sequence. If at the end of
the sequence all blocks for it is still clean, then the sequence block
will be stored.

Signed-off-by: eopXD <yuehtingc@nvidia.com>

eopXD commented on cpp/include/tensorrt_llm/batch_manager/kvCacheManager.h:
```diff
@@ -745,7 +759,8 @@ class WindowBlockManager
     //! \param blockKeys Key of each block.
     //! \param blockIds Id of each block.
-    void storeBlocks(std::vector<BlockKey> const& blockKeys, std::vector<KVCacheBlock::IdType> const& blockIds);
+    //! \return Number of actual blocks stored.
+    SizeType32 storeBlocks(std::vector<BlockKey> const& blockKeys, std::vector<KVCacheBlock::IdType> const& blockIds);
```
I would argue that BlockManager::storeBlocks doesn't even need to exist. However, that is a non-functional refactor and doesn't directly related to the scope of change of this merge request. Let me keep this in my note and mark it as resolve for now.

eopXD commented on cpp/tensorrt_llm/batch_manager/trtGptModelInflightBatching.cpp:
```diff
@@ -352,15 +352,13 @@ TrtGptModelInflightBatching::TrtGptModelInflightBatching(std::shared_ptr<nvinfer
         auto const numKvHeads = mModelConfig.getNbKvHeads(0);
 
-        mRewindInputs = RewindInputs{maxBlocksPerSeq, isUseOneMoreBlock, numKvHeads};
+        mRewindInputs = RewindInputs{maxBlocksPerSeq, /*isUseOneMoreBlock*/ false, numKvHeads};
```
Unrelated to this MR. Marking as resolved.

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -2394,9 +2597,8 @@ void KVCacheManager::removeToken(RequestIdType requestId)
-        SizeType32 const maxTokensInWindow = metadata.maxTokenNum;
-        SizeType32 const tokensInWindow = sequence.getNumTokens() % maxTokensInWindow;
-        if (tokensInWindow % getTokensPerBlock() == 0 && tokensInWindow <= maxTokensInWindow)
+        SizeType32 const tokensInWindow = sequence.getNumTokens() % windowSize;
+        if (tokensInWindow % getTokensPerBlock() == 0)
```
Bad suggestion and unrelated to scope of change of this MR, marking as resolved.

tensorrt-cicd commented:
[PR_Github #19779](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19779/) [ run ] triggered by Bot

eopXD commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #19916](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19916/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19916](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19916/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #14992](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14992/) completed with status: 'FAILURE'

eopXD committed:
Remove restriction to release a non-leaf block in the eviction policy

Signed-off-by: eopXD <yuehtingc@nvidia.com>

eopXD committed:
Fix WindowBlockManager::freeChildren

The function is calling EvictionPolicy::claimBlock which removes the
block from the free queue held by the EvictionPolicy object. Calling
upon a non-leaf block will result in the whole branch be 'claimed'
and not reusable,

This was not correct for full attention too but the manager acted
correctly because we call releaseBlocks at the end of a sequence
and releaseBlock will add the blocks back to the free queue.

This commit fixes it by only claiming the block we are are going to
acquire through WindowBlockManager::getFreeBlock.

Signed-off-by: eopXD <yuehtingc@nvidia.com>

eopXD commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #20510](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20510/) [ run ] triggered by Bot

eopXD committed:
Fix secondary block offload mechanism and update related unit test

Upon correcting WindowBlockManager::freeChildren, the existing bug for block offload
to secondary block pool under WindowBlockManager::freeChildren is exposed.

We should call claim upon the block 'after' the offload, not before.
This commit fixes it.

Unit test under kvCacheManagerTest.cpp is updated accordingly.
Added detail comments to illustrate the actual behaviors.

Signed-off-by: eopXD <yuehtingc@nvidia.com>

eopXD commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #20575](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20575/) [ run ] triggered by Bot

eopXD commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #20620](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20620/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #15568](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15568/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #20629](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20629/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20629](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20629/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15576](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15576/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #20731](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20731/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15665](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15665/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/15665/test-results/rerun_report.html) for details.

eopXD committed:
Remove `isEnableBlockReuse` parameter from adjustBlocksIfNeeded and detach frontBlock

The functions deal with block detach and block storage action only
happens when a sequence is removed.

Signed-off-by: eopXD <yuehtingc@nvidia.com>

thorjohnsen reviewed (approved):
Thank you for addressing my last review concerns. I have no further concerns with this PR.

2ez4bz reviewed (approved):
Approving; however, I am not sure why my review was needed in the first place, as I see no changes to any torch modeling files in the PR.
